src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -1972,15 +1972,170 @@ public void observe(@NonNull SCMHead head, @NonNull SCMRevision revision) throws
+                throw new IllegalStateException(
+                        "Name of created project " + project + " did not match expected " + encodedName);
+            }
+            // HACK ALERT
+            // ==========
+            // We don't want to trigger a save, so we will do some trickery to ensure that observer.created(project)
+            // performs the save
+            BulkChange bc = new BulkChange(project);
+            try {
+                if (project.getDisplayNameOrNull() == null && !rawName.equals(encodedName)) {
+                    project.setDisplayName(rawName);
+                }
+            } catch (IOException e) {
+                // ignore even if it does happen we didn't want a save
+            } finally {
+                bc.abort();
+            }
+            // decorate contract is to ensure it dowes not trigger a save
+            _factory.decorate(project);
+            // ok it is now up to the observer to ensure it does the actual save.
+            observer.created(project);
+            doAutomaticBuilds(head, revision, rawName, project, null, null);
+        }
+
+        private boolean isChangesDetected(@NonNull SCMRevision revision, @NonNull P project, SCMRevision scmLastBuiltRevision) {
+            boolean changesDetected = false;
+
+            if (revision.isDeterministic()) {
+                if (!revision.equals(scmLastBuiltRevision)) {
+                    changesDetected = true;
+                }
+            } else {
+                // TODO check if we should compare the revisions anyway...
+                // the revisions may not be deterministic, as in two checkouts of the same revision
+                // may result in additional changes (looking at you CVS) but that doesn't mean
+                // that the revisions are not capable of checking for differences, e.g. if
+                // rev1 == 2017-12-13 and rev2 == 2017-12-12 these could show as non-equal while both
+                // being non-deterministic (because they just specify the day not the timestamp within
+                // the day.
+
+                // fall back to polling when we have a non-deterministic revision/hash.
+                SCMTriggerItem scmProject = SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(project);
+                if (scmProject != null) {
+                    PollingResult pollingResult = scmProject.poll(listener);
+                    if (pollingResult.hasChanges()) {
+                        changesDetected = true;
+                    }
+                }
+            }
+            return changesDetected;
+        }
+
+        private @NonNull Action[] getRevisionActions(@NonNull SCMRevision revision, String rawName) {
+            Action[] revisionActions = new Action[0];
+            try {
+                List<Action> actions = source.fetchActions(revision, event, listener);
+                revisionActions = actions.toArray(new Action[0]);
```

Comment:
Slight performance enhancement...
```suggestion
                revisionActions = actions.toArray(new Action[actions.size()]);
```